pandas/_libs/tslibs/period.pyx

```diff
@@ -129,6 +146,285 @@ cdef char* c_strftime(date_info *dinfo, char *fmt):
+
+    absdays = absdate_from_ymd(year, month, day)
+    unix_date = absdays - ORD_OFFSET
+
+    if freq >= FR_SEC:
+        seconds = unix_date * 86400 + hour * 3600 + minute * 60 + second
+
+        if freq == FR_MS:
+            return seconds * 1000 + microseconds / 1000
+
+        elif freq == FR_US:
+            return seconds * 1000000 + microseconds
+
+        elif freq == FR_NS:
+            return (seconds * 1000000000 +
+                    microseconds * 1000 + picoseconds / 1000)
+
+        else:
+            return seconds
+
+    elif freq == FR_MIN:
+        return unix_date * 1440 + hour * 60 + minute
+
+    elif freq == FR_HR:
+        return unix_date * 24 + hour
+
+    elif freq == FR_DAY:
+        return unix_date
+
+    elif freq == FR_UND:
+        return unix_date
+
+    elif freq == FR_BUS:
+        # calculate the current week assuming sunday as last day of a week
+        # Jan 1 0001 is a Monday, so subtract 1 to get to end-of-week
+        weeks = (unix_date + ORD_OFFSET - 1) / 7
+        # calculate the current weekday (in range 1 .. 7)
+        delta = (unix_date + ORD_OFFSET - 1) % 7 + 1
+        # return the number of business days in full weeks plus the business
+        # days in the last - possible partial - week
+        if delta <= 5:
+            return (weeks * 5) + delta - BDAY_OFFSET
+        else:
+            return (weeks * 5) + (5 + 1) - BDAY_OFFSET
+
+    elif freq_group == FR_WK:
+        day_adj = freq - FR_WK
+        return (unix_date + ORD_OFFSET - (1 + day_adj)) / 7 + 1 - WEEK_OFFSET
+
+    # raise ValueError
+
+
+cdef int get_date_info(int64_t ordinal, int freq, date_info *dinfo) nogil:
+    cdef:
+        int64_t absdate
+        double abstime
+
+    absdate = get_python_ordinal(ordinal, freq);
+    abstime = get_abs_time(freq, absdate - ORD_OFFSET, ordinal)
+
+    while abstime < 0:
+        abstime += 86400
+        absdate -= 1
+
+    while abstime >= 86400:
+        abstime -= 86400
+        absdate += 1
+
+    dInfoCalc_SetFromAbsDateTime(dinfo, absdate, abstime)
+    return 0
+
+
+cdef int64_t get_python_ordinal(int64_t period_ordinal, int freq) nogil:
+    """
+    Returns the proleptic Gregorian ordinal of the date, as an integer.
+    This corresponds to the number of days since Jan., 1st, 1AD.
+    When the instance has a frequency less than daily, the proleptic date
+    is calculated for the last day of the period.
+    """
+    cdef:
+        asfreq_info af_info
+        freq_conv_func toDaily = NULL
+
+    if freq == FR_DAY:
+        return period_ordinal + ORD_OFFSET
+
+    toDaily = get_asfreq_func(freq, FR_DAY)
+    get_asfreq_info(freq, FR_DAY, 'E', &af_info)
+    return toDaily(period_ordinal, &af_info) + ORD_OFFSET
+
+
+cdef int dInfoCalc_SetFromAbsDateTime(date_info *dinfo,
+                                      int64_t absdate, double abstime) nogil:
+    """
+    Set the instance's value using the given date and time.
+    Assumes GREGORIAN_CALENDAR.
```

Comment:
ideally add some full doc-strings on these which list the args and types

Reply:
Will do.  Mind doing those in the next round to avoid clogging the CI?

Reply:
that's fine